lib/internal/process/esm_loader.js

```diff
@@ -24,12 +24,17 @@ function initializeImportMetaObject(wrap, meta) {
   }
 }
 
-async function importModuleDynamicallyCallback(wrap, specifier) {
-  if (callbackMap.has(wrap)) {
-    const { importModuleDynamically } = callbackMap.get(wrap);
+async function importModuleDynamicallyCallback(referrer, specifier) {
+  // compiled by vm.compileFunction
+  if (typeof referrer === 'string') {
```

Comment:
I very specifically made sure everything goes through the map in order to hit dynamic import. Please register the functions in the map instead.

Reply:
@devsnek dynamic import was implemented here in the commit at https://github.com/nodejs/node/pull/21573/commits/7a814eef9992028dcb36ce11ced7b009f5004749, but there was a bug due to the lifetime of the function not being available. I really hope we can fix this, and it would be great to get the support going for this here, but we shouldn't let an experimental feature block this PR when it is a completely unrelated issue.

Reply:
To be clear, making the compiled function lifetime persistent is a memory leak, and the implementation of importModuleDynamically seems naturally leaking to me, so we need to resolve those architectural issues first. I'm really keen to discuss further in the issue about this specifically.

Reply:
I don't understand why support for `importDynamically` in `vm.compileFunction` has to be added in this PR. Previously it simply does not work and throws and error, and nobody complains, so I don't think anyone is hitting that path. We can just use some hidden argument in the binding only for internal uses (that is, only define the host definitions when the binding is invoked by CJS compiler) and let the user land `vm.compileFunction` continue to throw.

Reply:
(my previous comment seems to have vanished, so reposting)

I've implemented exactly as proposed by @joyeecheung here now.

@devsnek I've also added a test that `vm.compileFunction` will throw the unhandled callback issue for any usage of dynamic import. This means this PR now makes no changes to the VM API at all.

Reply:
that seems fine to me, this definitely seems like something to tackle in another pr.

Reply:
this still needs to be removed, as well as the changes in module_wrap.cc